src/appleseed/foundation/platform/x86timer.cpp

```diff
@@ -102,13 +107,67 @@ uint64 X86Timer::read()
 
     uint32 h, l;
 
-    asm volatile
-    ("  rdtsc;                  \
-        mov %%edx, %0;          \
-        mov %%eax, %1"
-        : "=rm" (h), "=rm" (l)  // outputs
-        :                       // inputs
-        : "%eax", "%edx"        // clobbered registers
+    asm volatile (
+        "cpuid\n\t"                         // force in-order execution of the RDTSC instruction
+        "rdtsc\n\t"
+        "mov %%edx, %0\n\t"
+        "mov %%eax, %1\n\t"
+        : "=r" (h), "=r" (l)                // outputs
+        :                                   // inputs
+        : "%rax", "%rbx", "%rcx", "%rdx"    // clobbered registers
+    );
+
+    return (static_cast<uint64>(h) << 32) | l;
+
+// Other platforms.
+#else
+
+    #error The x86 timer is not supported on this platform.
+
+#endif
+}
+
+uint64 X86Timer::read_end()
+{
+// Visual C++, 32-bit mode.
+#if defined _MSC_VER && !defined _WIN64
+
+    uint32 h, l;
+
+    __asm
+    {
+        rdtscp
+        mov h, edx
+        mov l, eax
+        cpuid                               // prevent instructions coming afterward from executing before the RDTSCP instruction
+    }
+
+    return (static_cast<uint64>(h) << 32) | l;
+
+// Visual C++, 64-bit mode.
+#elif defined _MSC_VER && defined _WIN64
+
+    unsigned int aux;
+    const uint64 value = __rdtscp(&aux);
+
+    int cpu_info[4];
+    __cpuid(cpu_info, 0);                   // prevent instructions coming afterward from executing before the RDTSCP instruction
+
+    return value;
+
+// gcc.
+#elif defined __GNUC__
+
+    uint32 h, l;
+
+    asm volatile (
+        "rdtscp\n\t"
+        "mov %%edx, %0\n\t"
+        "mov %%eax, %1\n\t"
+        "cpuid\n\t"                         // prevent instructions coming afterward from executing before the RDTSCP instruction
+        : "=r" (h), "=r" (l)                // outputs
+        :                                   // inputs
+        : "%rax", "%rbx", "%rcx", "%rdx"    // clobbered registers
```

Comment:
For clobbered register list, you do not need to use `%`.  However, in correct CPUID implementation, you wouldn't have to declare clobbered list explicitly in the first place.